src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+
+-- convert from epoch related time to Rata Die related
+local function local_rd(secs)
+    return math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY)
+end
+
+-- convert UTC seconds to local seconds, adjusting by timezone
+local function local_secs(obj)
+    return obj.epoch + obj.tzoffset * 60
+end
+
+local function utc_secs(epoch, tzoffset)
+    return epoch - tzoffset * 60
+end
+
+-- get epoch seconds, shift to the local timezone
+-- adjust from 1970-related to 0000-related time
+-- then return dt in those coordinates (number of days
+-- since Rata Die date)
+local function local_dt(obj)
+    return builtin.tnt_dt_from_rdn(local_rd(local_secs(obj)))
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 then
+        secs = secs - 1
+        nsec = nsec + NANOS_PER_SEC
+    elseif nsec >= NANOS_PER_SEC then
+        secs = secs + 1
+        nsec = nsec - NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_cmp(lhs, rhs)
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        return nil
+    end
+    local sdiff = lhs.epoch - rhs.epoch
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
+end
+
+local function datetime_eq(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc ~= nil and rc == 0
+end
+
+local function datetime_lt(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc < 0
+end
+
+local function datetime_le(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc <= 0
+end
+
+local function datetime_serialize(self)
+    return { epoch = self.epoch, nsec = self.nsec,
+             tzoffset = self.tzoffset, tzindex = 0 }
+end
+
+local parse_zone
+
+local function datetime_new_raw(epoch, nsec, tzoffset, tzindex)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset or 0
+    dt_obj.tzindex = tzindex or 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, fraction, offset)
+    local epochV = dt ~= nil and (builtin.tnt_dt_rdn(dt) - DT_EPOCH_1970_OFFSET) *
+                   SECS_PER_DAY or 0
+    local secsV = secs or 0
+    local fracV = fraction or 0
+    local ofsV = offset or 0
+    return datetime_new_raw(epochV + secsV - ofsV * 60, fracV, ofsV)
+end
+
+local function get_timezone(offset)
+    if type(offset) == 'number' then
+        return offset
+    elseif type(offset) == 'string' then
+        return parse_zone(offset)
+    end
+end
+
+-- create datetime given attribute values from obj
+local function datetime_new(obj)
+    if obj == nil or type(obj) ~= 'table' then
+        return datetime_new_raw(0, 0, 0)
+    end
+
+    local ymd = false
+    local hms = false
+    local dt = DT_EPOCH_1970_OFFSET -- default is 1970-01-01
+
+    local y = obj.year
+    if y ~= nil then
+        check_range(y, {1, 9999}, 'year')
```

Comment:
What about BC dates?

Reply:
If we would remove this check then underlying implementation could handle BC dates very easily. This artificial constraints introduced to make it more compatible with future box index support, and SQL standard constraints.

[Could easily be removed though. But we need to still check somehow, that underlying double in `epoch` is still able to handle seconds since epoch precisely, without any rounding]